File: app/src/main/java/sourcecontrol/utils/crypto/HashUtils.java
package sourcecontrol.utils.crypto;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Utility class for hashing data using SHA-1 algorithm.
 */
public class HashUtils {

    public static String sha1Hex(byte[] data) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-1");
            byte[] hash = digest.digest(data);
            return bytesToHex(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-1 algorithm not available", e);
        }
    }

    public static String sha1Hex(String data) {
        return sha1Hex(data.getBytes());
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }
}

File: src/lib/sourcecontrol/utils/hash-utils.ts
export class HashUtils {
  public static async sha1Hex(data: Uint8Array | string): Promise<string> {
    let bytes: Uint8Array;

    if (typeof data === "string") {
      bytes = new TextEncoder().encode(data);
    } else {
      bytes = data;
    }

    try {
      const hashBuffer = await crypto.subtle.digest(
        "SHA-1",
        new Uint8Array(bytes)
      );
      return this.bytesToHex(new Uint8Array(hashBuffer));
    } catch (error) {
      throw new Error(`SHA-1 algorithm not available: ${error}`);
    }
  }

  private static bytesToHex(bytes: Uint8Array): string {
    return Array.from(bytes)
      .map((byte) => byte.toString(16).padStart(2, "0"))
      .join("");
  }
}
